kernel/src/fs/configfs/tsm.rs

```diff
@@ -0,0 +1,206 @@
+            let fields = BranchNodeFields::new(name, attrs, weak_self.clone());
+            fields
+                .add_child(ReportSystem::new())
+                .expect("Failed to add report directory");
+
+            Tsm { fields }
+        })
+    }
+}
+
+inherit_sys_branch_node!(Tsm, fields, {
+    fn perms(&self) -> SysPerms {
+        SysPerms::DEFAULT_RW_PERMS
+    }
+});
+
+#[derive(Debug)]
+struct ReportSystem {
+    fields: BranchNodeFields<Report, Self>,
+}
+
+#[inherit_methods(from = "self.fields")]
+impl ReportSystem {
+    fn new() -> Arc<Self> {
+        let name = SysStr::from("report");
+        let builder = SysAttrSetBuilder::new();
+
+        let attrs = builder.build().expect("Failed to build attribute set");
+
+        Arc::new_cyclic(|weak_self| ReportSystem {
+            fields: BranchNodeFields::new(name, attrs, weak_self.clone()),
+        })
+    }
+
+    fn add_child(&self, new_child: Arc<Report>) -> Result<()>;
+}
+
+inherit_sys_branch_node!(ReportSystem, fields, {
+    fn perms(&self) -> SysPerms {
+        SysPerms::DEFAULT_RW_PERMS
+    }
+
+    fn create_child(&self, name: &str) -> Result<Arc<dyn SysObj>> {
+        let report = Report::new(SysStr::from(name.to_string()));
+        self.add_child(report.clone())?;
+        Ok(report)
+    }
+});
+
+#[derive(Debug)]
+pub struct TsmProvider {
+    provider: SysStr,
+    inblob: [u8; TSM_INBLOB_MAX],
+    inblob_len: usize,
+    outblob: Option<Box<[u8]>>,
+    generation: u32,
+}
+
+impl TsmProvider {
+    pub fn new() -> Self {
+        TsmProvider {
+            provider: SysStr::from("tdx_guest"),
+            inblob: [0u8; TSM_INBLOB_MAX],
+            inblob_len: 0,
+            outblob: None,
+            generation: 1,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Report {
+    fields: NormalNodeFields<Self>,
+    data: RwLock<TsmProvider>,
+}
+
+impl Report {
+    fn new(name: SysStr) -> Arc<Self> {
+        let mut builder = SysAttrSetBuilder::new();
+        builder.add(SysStr::from("inblob"), SysPerms::DEFAULT_RW_ATTR_PERMS);
+        builder.add(SysStr::from("outblob"), SysPerms::DEFAULT_RO_ATTR_PERMS);
+        builder.add(SysStr::from("provider"), SysPerms::DEFAULT_RO_ATTR_PERMS);
+        builder.add(SysStr::from("generation"), SysPerms::DEFAULT_RO_ATTR_PERMS);
+        let attrs = builder.build().expect("Failed to build attribute set");
+
+        Arc::new_cyclic(|weak_self| {
+            let fields = NormalNodeFields::new(name, attrs, weak_self.clone());
+
+            Report {
+                fields,
+                data: RwLock::new(TsmProvider::new()),
+            }
+        })
+    }
+}
+
+inherit_sys_leaf_node!(Report, fields, {
+    fn perms(&self) -> SysPerms {
+        SysPerms::DEFAULT_RW_PERMS
+    }
+
+    fn read_attr_at(&self, name: &str, offset: usize, writer: &mut VmWriter) -> Result<usize> {
+        match name {
+            "provider" => {
+                let mut printer = VmPrinter::new_skip(writer, offset);
```

Comment:
For all attribute read/write operations, always acquire the lock first to make it apparent that the code is written with concurrent access in mind.

```rust
        match name {
            "provider" => {
                let data = self.data.read();
                ...
            }
            "outblob" => {
                let mut data = self.data.write();
                ...          
            }
            "generation" => {
                let data = self.data.read();
                ...                
            }
        }
```